Before we start anything. We need to download the libarries that we plan on using.

In [161]:
pip install pandas seaborn

Note: you may need to restart the kernel to use updated packages.


In [162]:
import pandas as pd
import seaborn as sns

Now that all of our libaries are downloaded and imported, we can start importing our data

For this lab we are going to be working with an excel spreadsheet of trending youtube videos in the United States 


## Importing data
before we can do anything with the data, we must but it into a pandas dataFrame so that we can easily view and manipulate the data.

Since the data is in an CSV (Comma Seperated Values) document, we will use `pd.read_csv()`

In [163]:
# Import data
youtube_data = pd.read_csv('USvideos.csv')
youtube_data

video_id trending_date  \
0      2kyS6SvSYSE      17.14.11   
1      1ZAPwfrtAFY      17.14.11   
2      5qpjK5DgCt4      17.14.11   
3      puqaWrEC7tY      17.14.11   
4      d380meD0W0M      17.14.11   
...            ...           ...   
40944  BZt0qjTWNhw      18.14.06   
40945  1h7KV2sjUWY      18.14.06   
40946  D6Oy4LfoqsU      18.14.06   
40947  oV0zkMe1K8s      18.14.06   
40948  ooyjaVdt-jA      18.14.06   

                                                   title  \
0                     WE WANT TO TALK ABOUT OUR MARRIAGE   
1      The Trump Presidency: Last Week Tonight with J...   
2      Racist Superman | Rudy Mancuso, King Bach & Le...   
3                       Nickelback Lyrics: Real or Fake?   
4                               I Dare You: GOING BALD!?   
...                                                  ...   
40944                       The Cat Who Caught the Laser   
40945                         True Facts : Ant Mutualism   
40946  I GAVE SAFIYA NYGAARD A PERFECT HAIR MAKEOVER ...   
40947                How Black Panther Should Have Ended   
40948  Official Call of Duty®: Black Ops 4 — Multipla...   

                  channel_title  category_id              publish_time  \
0                  CaseyNeistat           22  2017-11-13T17:13:01.000Z   
1               LastWeekTonight           24  2017-11-13T07:30:00.000Z   
2                  Rudy Mancuso           23  2017-11-12T19:05:24.000Z   
3         Good Mythical Morning           24  2017-11-13T11:00:04.000Z   
4                      nigahiga           24  2017-11-12T18:01:41.000Z   
...                         ...          ...                       ...   
40944             AaronsAnimals           15  2018-05-18T13:00:04.000Z   
40945                  zefrank1           22  2018-05-18T01:00:06.000Z   
40946                Brad Mondo           24  2018-05-18T17:34:22.000Z   
40947  How It Should Have Ended            1  2018-05-17T17:00:04.000Z   
40948              Call of Duty           20  2018-05-17T17:09:38.000Z   

                                                    tags     views   likes  \
0                                        SHANtell martin    748374   57527   
1      last week tonight trump presidency|"last week ...   2418783   97185   
2      racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434  146033   
3      rhett and link|"gmm"|"good mythical morning"|"...    343168   10172   
4      ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095731  132235   
...                                                  ...       ...     ...   
40944  aarons animals|"aarons"|"animals"|"cat"|"cats"...   1685609   38160   
40945                                             [none]   1064798   60008   
40946  I gave safiya nygaard a perfect hair makeover ...   1066451   48068   
40947  Black Panther|"HISHE"|"Marvel"|"Infinity War"|...   5660813  192957   
40948      call of duty|"cod"|"activision"|"Black Ops 4"  10306119  357079   

       dislikes  comment_count  \
0          2966          15954   
1          6146          12703   
2          5339           8181   
3           666           2146   
4          1989          17518   
...         ...            ...   
40944      1385           2657   
40945       382           3936   
40946      1032           3992   
40947      2846          13088   
40948    212976         144795   

                                       thumbnail_link  comments_disabled  \
0      https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg              False   
1      https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg              False   
2      https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg              False   
3      https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg              False   
4      https://i.ytimg.com/vi/d380meD0W0M/default.jpg              False   
...                                               ...                ...   
40944  https://i.ytimg.com/vi/BZt0qjTWNhw/default.jpg              False   
40945  https:

Data analysis always start with a question we want to solve. 

For this workshop, we are going to try to answer the following questions
- who has made the most trending videos?
- what are the most common words in a trending title?
- what is the best time to post a video if we want it to be trending?

However, before we can start answering any of these quesitons, we are going to need gain a better understanding of our data through a process called <b>data exploration<b/>

## Exploring Our Data

In [164]:
youtube_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40949 entries, 0 to 40948
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   video_id                40949 non-null  object
 1   trending_date           40949 non-null  object
 2   title                   40949 non-null  object
 3   channel_title           40949 non-null  object
 4   category_id             40949 non-null  int64 
 5   publish_time            40949 non-null  object
 6   tags                    40949 non-null  object
 7   views                   40949 non-null  int64 
 8   likes                   40949 non-null  int64 
 9   dislikes                40949 non-null  int64 
 10  comment_count           40949 non-null  int64 
 11  thumbnail_link          40949 non-null  object
 12  comments_disabled       40949 non-null  bool  
 13  ratings_disabled        40949 non-null  bool  
 14  video_error_or_removed  40949 non-null  bool  
 15  de

Describe integers

In [165]:
youtube_data.describe()

category_id         views         likes      dislikes  comment_count
count  40949.000000  4.094900e+04  4.094900e+04  4.094900e+04   4.094900e+04
mean      19.972429  2.360785e+06  7.426670e+04  3.711401e+03   8.446804e+03
std        7.568327  7.394114e+06  2.288853e+05  2.902971e+04   3.743049e+04
min        1.000000  5.490000e+02  0.000000e+00  0.000000e+00   0.000000e+00
25%       17.000000  2.423290e+05  5.424000e+03  2.020000e+02   6.140000e+02
50%       24.000000  6.818610e+05  1.809100e+04  6.310000e+02   1.856000e+03
75%       25.000000  1.823157e+06  5.541700e+04  1.938000e+03   5.755000e+03
max       43.000000  2.252119e+08  5.613827e+06  1.674420e+06   1.361580e+06

Check duplicates

In [166]:
has_duplicates = youtube_data.duplicated()
has_duplicates.sum()

48

In [167]:
youtube_data[has_duplicates].sort_values(by=['title']).head()

video_id trending_date  \
34919  3g5O-kT9m8k      18.15.05   
34905  jzLlsbdrwQk      18.15.05   
34923  D2mxKEa2xmA      18.15.05   
34913  4oqvNR1o3Zo      18.15.05   
34914  96oKlWv5wSo      18.15.05   

                                                   title       channel_title  \
34919                       $1000 Survival Kit in a Case  CrazyRussianHacker   
34905                           $17 Pet vs. $100,000 Pet        BuzzFeedBlue   
34923       10 Most DEVASTATING Cyber Attacks in History      MatthewSantoro   
34913  24 HOUR BOX FORT PRISON ESCAPE ROOM!! 📦🚔 Diggi...           Papa Jake   
34914  AMAZING All You Can Eat  BBQ KEBAB Buffet in I...   Strictly Dumpling   

       category_id              publish_time  \
34919           28  2018-05-13T18:51:13.000Z   
34905           22  2018-05-13T15:00:57.000Z   
34923           24  2018-05-13T13:00:01.000Z   
34913           24  2018-05-13T16:42:33.000Z   
34914           19  2018-05-13T20:17:14.000Z   

                                                    tags    views  likes  \
34919  $1000 Survival Kit|"Survival Kit in a Case"|"p...  1289899  51318   
34905  buzzfeed|"worth it"|"cheap vs. expensive"|"ste...  3081033  60379   
34923  computer hacks|"cyber attack"|"computer virus"...   158406   6773   
34913  box fort|"fort"|"kids"|"funny"|"fun"|"no swear...  1030616  20251   
34914  indian buffet|"kebab buffet"|"buffet"|"all you...   456475  10561   

       dislikes  comment_count  \
34919      1359           4332   
34905      6857           7796   
34923       364            947   
34913      3290           3567   
34914       337           1384   

                                       thumbnail_link  comments_disabled  \
34919  https://i.ytimg.com/vi/3g5O-kT9m8k/default.jpg              False   
34905  https://i.ytimg.com/vi/jzLlsbdrwQk/default.jpg              False   
34923  https://i.ytimg.com/vi/D2mxKEa2xmA/default.jpg              False   
34913  https://i.ytimg.com/vi/4oqvNR1o3Zo/default.jpg              False   
34914  https://i.ytimg.com/vi/96oKlWv5wSo/default.jpg              False   

       ratings_disabled  video_error_or_removed  \
34919             False                   False   
34905             False                   False   
34923             False                   False   
34913             False                   False   
34914             False                   False   

                                             description  
34919  $15 Survival Kit Unboxing - https://youtu.be/0...  
34905  “Who is Lennox and why is he so special?”Credi...  
34923  Download the best VPN on the market, NORDVPN n...  
34913  In Todays 24 hour Box Fort video papa Jake bui...  
34914  I'm always on the lookout for new amazing buff...

This means that there are 48 rows that have all the same values as some other row in the dateset.

since we don't need them for our analysis, we can get rid of them with using `df.drop_duplicates()`

In [168]:
youtube_data = youtube_data.drop_duplicates()
youtube_data.duplicated().sum()

0

## Who has made the most trending videos

to solve this particular question we are going to need a list of all the videos.

However, we want to make sure that every video is only showing up once.

we are going to use `df.duplicated` again but this time we will list the columns we want to check for duplicates on and then use that list as a parameter 

In [169]:
cols_to_check_unique = ['title', 'channel_title']
youtube_data.duplicated(cols_to_check_unique).sum()

34433

We can now see that there are 34433 rows that share a title and a channel title with some other row in the dataset.

Once again we are going to drop these rows using `df.drop_duplicates()` but we will pass our list of columns to it like we did `df.duplicated()`

*Be aware that `df.drop_duplicates()` will keep the first row it finds and delete any duplicates afterwards. If you want to keep the last ones, use `df.drop_duplicates(COL_TO_CHECK,keep='last')`*

In [170]:
# Drop duplicates while keeping the most recent appearance
unique_video_data = youtube_data.drop_duplicates(cols_to_check_unique)

Lets count the duplicates again to make sure that we got rid of everything that we wanted to

In [171]:
unique_video_data.duplicated(cols_to_check_unique).sum()

0

If the value is zero than we have set everything up corrent and can now continue with the analysis

Next up, we need to put all of our rows of data into groups based on the channel that made the video.

We can do this by using `df.groupby(by=['channel_title'])` and then use `.size()` on the result to get the number of rows in each group

In [172]:
channels = unique_video_data.groupby(by=['channel_title']).size()
channels.head(10)

channel_title
12 News                                               1
1MILLION Dance Studio                                 4
1theK (원더케이)                                          5
20th Century Fox                                     17
2CELLOS                                               1
3Blue1Brown                                           3
3D Printing Nerd                                      2
42Fab - Metalworking and Multi-Medium Fabrication     1
494ta                                                 1
4given4ever1                                          1
dtype: int64

Finally, we just need to sort these groups using `.sort_values()` make sure to set the ascending parameter to false

In [173]:
channels_by_trending_count = channels.sort_values(ascending=False)
channels_by_trending_count.head()

channel_title
ESPN                                      84
TheEllenShow                              74
The Tonight Show Starring Jimmy Fallon    72
Jimmy Kimmel Live                         70
The Late Show with Stephen Colbert        58
dtype: int64

## what are the most common words in trending titles?

In [174]:
titles = youtube_data.drop_duplicates(['title'])['title'].to_frame()
titles

title
0                     WE WANT TO TALK ABOUT OUR MARRIAGE
1      The Trump Presidency: Last Week Tonight with J...
2      Racist Superman | Rudy Mancuso, King Bach & Le...
3                       Nickelback Lyrics: Real or Fake?
4                               I Dare You: GOING BALD!?
...                                                  ...
40760                      CRAYOLA MAKEUP | HIT OR MISS?
40761  First Take reacts: Kyrie Irving says contract ...
40762  Drop the Mic w/ Ashton Kutcher & Sean Diddy Combs
40764  Carla Makes BA Smashburgers | From the Test Ki...
40766  Katherine Langford on 13 Reasons Why, Australi...

[6455 rows x 1 columns]

In [175]:
import re
def create_word_cols(row):
    words = re.findall(r'\w+', row['title'].lower())
    word_count = {}
    for word in words:
        word_count[word] = words.count(word)
    return word_count

In [176]:
titles['word count'] = titles.apply(create_word_cols, axis=1)
word_count_df = pd.json_normalize(titles['word count'])
word_count_df

we  want   to  talk  about  our  marriage  the  trump  presidency  ...  \
0     1.0   1.0  1.0   1.0    1.0  1.0       1.0  NaN    NaN         NaN  ...   
1     NaN   NaN  NaN   NaN    NaN  NaN       NaN  1.0    1.0         1.0  ...   
2     NaN   NaN  NaN   NaN    NaN  NaN       NaN  NaN    NaN         NaN  ...   
3     NaN   NaN  NaN   NaN    NaN  NaN       NaN  NaN    NaN         NaN  ...   
4     NaN   NaN  NaN   NaN    NaN  NaN       NaN  NaN    NaN         NaN  ...   
...   ...   ...  ...   ...    ...  ...       ...  ...    ...         ...  ...   
6450  NaN   NaN  NaN   NaN    NaN  NaN       NaN  NaN    NaN         NaN  ...   
6451  NaN   NaN  NaN   NaN    NaN  NaN       NaN  NaN    NaN         NaN  ...   
6452  NaN   NaN  NaN   NaN    NaN  NaN       NaN  1.0    NaN         NaN  ...   
6453  NaN   NaN  NaN   NaN    NaN  NaN       NaN  1.0    NaN         NaN  ...   
6454  NaN   NaN  NaN   NaN    NaN  NaN       NaN  NaN    NaN         NaN  ...   

      ratatouille  nun  crayola  contract  extension  sense   ba  \
0             NaN  NaN      NaN       NaN        NaN    NaN  NaN   
1             NaN  NaN      NaN       NaN        NaN    NaN  NaN   
2             NaN  NaN      NaN       NaN        NaN    NaN  NaN   
3             NaN  NaN      NaN       NaN        NaN    NaN  NaN   
4             NaN  NaN      NaN       NaN        NaN    NaN  NaN   
...           ...  ...      ...       ...        ...    ...  ...   
6450          NaN  NaN      1.0       NaN        NaN    NaN  NaN   
6451          NaN  NaN      NaN       1.0        1.0    1.0  NaN   
6452          NaN  NaN      NaN       NaN        NaN    NaN  NaN   
6453          NaN  NaN      NaN       NaN        NaN    NaN  1.0   
6454          NaN  NaN      NaN       NaN        NaN    NaN  NaN   

      smashburgers  katherine  langford  
0              NaN        NaN       NaN  
1              NaN        NaN       NaN  
2              NaN        NaN       NaN  
3              NaN        NaN       NaN  
4              NaN        NaN       NaN  
...            ...        ...       ...  
6450           NaN        NaN       NaN  
6451           NaN        NaN       NaN  
6452           NaN        NaN       NaN  
6453           1.0        NaN       NaN  
6454           NaN        1.0       1.0  

[6455 rows x 10336 columns]

In [177]:
word_count_df = word_count_df.fillna(0)
word_count_df

we  want   to  talk  about  our  marriage  the  trump  presidency  ...  \
0     1.0   1.0  1.0   1.0    1.0  1.0       1.0  0.0    0.0         0.0  ...   
1     0.0   0.0  0.0   0.0    0.0  0.0       0.0  1.0    1.0         1.0  ...   
2     0.0   0.0  0.0   0.0    0.0  0.0       0.0  0.0    0.0         0.0  ...   
3     0.0   0.0  0.0   0.0    0.0  0.0       0.0  0.0    0.0         0.0  ...   
4     0.0   0.0  0.0   0.0    0.0  0.0       0.0  0.0    0.0         0.0  ...   
...   ...   ...  ...   ...    ...  ...       ...  ...    ...         ...  ...   
6450  0.0   0.0  0.0   0.0    0.0  0.0       0.0  0.0    0.0         0.0  ...   
6451  0.0   0.0  0.0   0.0    0.0  0.0       0.0  0.0    0.0         0.0  ...   
6452  0.0   0.0  0.0   0.0    0.0  0.0       0.0  1.0    0.0         0.0  ...   
6453  0.0   0.0  0.0   0.0    0.0  0.0       0.0  1.0    0.0         0.0  ...   
6454  0.0   0.0  0.0   0.0    0.0  0.0       0.0  0.0    0.0         0.0  ...   

      ratatouille  nun  crayola  contract  extension  sense   ba  \
0             0.0  0.0      0.0       0.0        0.0    0.0  0.0   
1             0.0  0.0      0.0       0.0        0.0    0.0  0.0   
2             0.0  0.0      0.0       0.0        0.0    0.0  0.0   
3             0.0  0.0      0.0       0.0        0.0    0.0  0.0   
4             0.0  0.0      0.0       0.0        0.0    0.0  0.0   
...           ...  ...      ...       ...        ...    ...  ...   
6450          0.0  0.0      1.0       0.0        0.0    0.0  0.0   
6451          0.0  0.0      0.0       1.0        1.0    1.0  0.0   
6452          0.0  0.0      0.0       0.0        0.0    0.0  0.0   
6453          0.0  0.0      0.0       0.0        0.0    0.0  1.0   
6454          0.0  0.0      0.0       0.0        0.0    0.0  0.0   

      smashburgers  katherine  langford  
0              0.0        0.0       0.0  
1              0.0        0.0       0.0  
2              0.0        0.0       0.0  
3              0.0        0.0       0.0  
4              0.0        0.0       0.0  
...            ...        ...       ...  
6450           0.0        0.0       0.0  
6451           0.0        0.0       0.0  
6452           0.0        0.0       0.0  
6453           1.0        0.0       0.0  
6454           0.0        1.0       1.0  

[6455 rows x 10336 columns]

In [178]:
word_count_df = word_count_df.transpose()
word_count_df

0     1     2     3     4     5     6     7     8     9     ...  \
we             1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
want           1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
to             1.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0  ...   
talk           1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
about          1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
...            ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
sense          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
ba             0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
smashburgers   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
katherine      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
langford       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   

              6445  6446  6447  6448  6449  6450  6451  6452  6453  6454  
we             0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
want           0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
to             0.0   0.0   1.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0  
talk           0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
about          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
...            ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
sense          0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0  
ba             0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0  
smashburgers   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0  
katherine      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0  
langford       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0  

[10336 rows x 6455 columns]

In [182]:
word_count_df['Total'] = word_count_df.apply(sum, axis=1)
word_count_df['Total'].sort_values(ascending=False).head(10)

the         6680.0
a           3128.0
s           3084.0
to          2812.0
in          2236.0
of          2176.0
official    1992.0
with        1924.0
on          1696.0
and         1636.0
Name: Total, dtype: float64

## What is the best time to post a video